include/turnstile.h

```diff
@@ -0,0 +1,40 @@
+#ifndef _SYS_TURNSTILE_H_
+#define _SYS_TURNSTILE_H_
+
+#include <common.h>
+
+typedef struct turnstile turnstile_t;
+
+/*! \brief Initializes turnstiles.
+ *
+ * \warning To be called only from early kernel initialization! */
+void turnstile_init(void);
+
+/*! \brief Allocates turnstile entry. */
+turnstile_t *turnstile_alloc(void);
+
+/*! \brief Deallocates turnstile entry. */
+void turnstile_destroy(turnstile_t *ts);
+
+/* Scheduler should call this after changing priority of a thread that is
+ * blocked on some turnstile.
+ *
+ * It will re-sort list of blocked threads, on which `td` is and possibly
+ * propagate `td`'s priority if it was increased.
+ *
+ * If `td` was lending priority and it was decreased, we don't change the
+ * lent priority. The borrowing thread should finish soon anyway and we
+ * don't care enough.
```

Comment:
@ilikeheaps maybe:
```
If `td` was lending priority and it was decreased, we don't unlend priority as unlending is expensive and the borrowing thread should finish soon.
```
? I think that `we don't care enough` is not appropriate.

Reply:
Done.